### ISSUE TITLE
Incorrect warning about wchar_t in C

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#155988]
Should this warn about `wchar_t` in C?

```c
#include <wchar.h>

int main (void)
{
  wchar_t foo;
  (void)foo;
  return 0;
}
```

```
% xcrun /Users/builder/llvm/llvm-install/bin/clang -fsyntax-only -Weverything ~/Desktop/test.c

/Users/builder/Desktop/test.c:5:3: warning: identifier 'wchar_t' conflicts with a C++ keyword [-Wc++-keyword]
    5 |   wchar_t foo;
      |   ^
```

_Originally posted by @seanm in https://github.com/llvm/llvm-project/issues/137234#issuecomment-3233883505_

**Comment by AaronBallman** [comment]
This does seem to require some pretty significant reworking to support correctly. We only have the `IdentifierInfo` but we need more context. The only idea I had (which may be a terrible one) would be to provide some sort of optional callback to the API so we can say "this might be reserved, but let's let the callback tell us if we're wrong".

But yeah, if we can't find a way to silence the diagnostic, it might make sense to turn it off for these identifiers, which would be unfortunate because we'd lose the diagnostic in a case like:
```
// C file, no headers included.
void wchar_t(void) {}
```
but not the end of the world given how easy it is to get transitive includes which would result in: https://godbolt.org/z/axq4K4ThK or similar.
